### PR TITLE
Correct issue 9509 missing index and column changes in utf8mb4 conversion

### PR DESCRIPTION
**Pull request by richard67**
Pull Request for Issue #9509 
 .
#### Summary of Changes

Add changing of columns and indexes to utf8(mb4) conversion scripts which have been forgotten before.
- Column `user_id` of table `#__user_keys` downsized to length of 150 so it fits to column `username` of table `#__users`.
- Changed column length of column `name` in table `#__users` to 400.
- Changed follwoing indexes (keys) to include column only with 1st 100 chars:
  - TABLE `#__menu`, KEY `idx_path`
  - TABLE `#__users`, KEY `idx_name`

Column `path` of table `#__menu` has already 1024 and was not enlarged in 3.5.0's joomla.sql, and so it is also not enlarged with this PR.

Note: This PR does not solve the problem that the utf8mb4 or utf8 combersion has to be executed again after an update e.g. from 3.5.0 to the 3.5.x in which this PR will be included. This can be manually done by the SQL statement provided in testing instructions below.

A solution for that has to be developed with another PR.
#### Testing Instructions

Step 1: Apply this patch on latest staging or a 3.5.0 final release, e.g. using the patchtester.

Step 2: In phpMyAdmin, execute the SQL statement from the new schema update script included in this PR:

> UPDATE `#__utf8_conversion` SET `converted` = 0;

The reason for this step is that this PR does not solve the problem that the utf8mb4 or utf8 combersion has to be executed again after an update e.g. from 3.5.0 to the 3.5.x in which this PR will be included, see description above.

Step 3: Go to "Extensions -> Manage -> Database".

Result: Database is not up to date, utf8mb4 (or utf8) conversion is shown as to be done.

Click the "Fix" button.

Result: Database up to date.

Check with phpMyadmin if the changes mentioned in the Summary of Changes above have been applied.

Result: Changes have been applied.

### CODE REVIEW

**Comment by ltraveler** [comment]
I've made update as you u said using custom url the update wasn't finished and at the and i've got a message like that:

*****

ERROR 504 - GATEWAY TIMEOUT

Why am I seeing this page?

The server that your request has reached is acting as a gateway or proxy to fulfil the request made by your client.

Web Browser => Web Front-End => Web Back-End

This server (Web Front-End) received an invalid response from an upstream (Web Back-End) server it accessed to fulfil the request.

In most cases this will not mean that the upstream server is down, but rather that the upstream server and the gateway/proxy do not agree on the protocol for exchanging data.

This problem is most commonly caused when there is a problem with IP communications between the Web Front and Back-Ends. Before you attempt to resolve this problem you should clear your browser cache completely.

Our support staff will be happy to assist you in resolving this issue. Please contact our Live Support or reply to any Tickets you may have received from our technicians for further assistance.

*****

So then I go to Joomla update again and the problem 
The Joomla! Core database tables have not been converted yet to UTF-8 Multibyte (utf8mb4).
Still exists and actually I do not know how to fix it...<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9510">issues.joomla.org/joomla-cms/9510</a>.</sub>

**Comment by ltraveler** [comment]
When I am clicking Fix in Joomla Upgrade after some waiting I am getting an error
405 Not Allowed

nginx/1.8.1

I increased in my php.ihi default_socket_timeout from 90 to 360
and in mysql.connect_timeout from 90 to 360
But it doesn't help. Please help me....<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9510">issues.joomla.org/joomla-cms/9510</a>.</sub>

**Comment by ltraveler** [comment]
The problem still exists:
The Joomla! Core database tables have not been converted yet to UTF-8 Multibyte (utf8mb4).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9510">issues.joomla.org/joomla-cms/9510</a>.</sub>

**Comment by wilsonge** [comment]
So this is definitely the fix we need - the problem is that if people upgrade via the file upload approach then we are in trouble cause the schema update (obviously) doesn't check upgrade or insert statements. meaning they won't run this query :(

**Comment by richard67** [comment]
@wilsonge "So this is definitely the fix we need" .. what do you mean with that? This PR? Or a fix for schema manager detecting inserts and updates, too (I would not recommend this then but maybe a special handling of updates on the utf8_conversion table)?

**Comment by wilsonge** [comment]
I mean it fixed the forum post issue that started all these problems :P

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on d2bb748c3c9bec3c5c675650089ea1ce610c5cf0<br /><br />used Method 1: For lazy testers for tests<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9510">issues.joomla.org/joomla-cms/9510</a>.</sub>

File: administrator/components/com_admin/sql/others/mysql/utf8mb4-conversion-02.sql [hunk 1]
```diff
@@ -49,6 +52,8 @@ ALTER TABLE `#__tags` MODIFY `alias` varchar(400) NOT NULL DEFAULT '';
 ALTER TABLE `#__ucm_content` MODIFY `core_type_alias` varchar(400) NOT NULL DEFAULT '' COMMENT 'FK to the content types table';
 ALTER TABLE `#__ucm_content` MODIFY `core_title` varchar(400) NOT NULL;
 ALTER TABLE `#__ucm_content` MODIFY `core_alias` varchar(400) NOT NULL DEFAULT '';
+ALTER TABLE `#__users` MODIFY `name` varchar(400) NOT NULL DEFAULT '';
+ALTER TABLE `#__user_keys` MODIFY `user_id` varchar(400) NOT NULL;
```
Comment (wilsonge) [inline administrator/components/com_admin/sql/others/mysql/utf8mb4-conversion-02.sql]: So this matches the username column which is limited to 150 characters - let's just fix this now and edit this in the joomla.sql and here to make everything "add up"

Comment (richard67) [inline administrator/components/com_admin/sql/others/mysql/utf8mb4-conversion-02.sql]: Sorry?

Comment (andrepereiradasilva) [inline administrator/components/com_admin/sql/others/mysql/utf8mb4-conversion-02.sql]: yeah joomla.sql needs the change in columns too.

Comment (richard67) [inline administrator/components/com_admin/sql/others/mysql/utf8mb4-conversion-02.sql]: It does not have these sizes yet?

Comment (andrepereiradasilva) [inline administrator/components/com_admin/sql/others/mysql/utf8mb4-conversion-02.sql]: oh, you're right sorry.

Comment (richard67) [inline administrator/components/com_admin/sql/others/mysql/utf8mb4-conversion-02.sql]: The sizes of the 2 columns above are already 400 in joomla.sql, and the indexes also are already limited to 100. That was the issue, to make the conversion script equal to Joomla!

Comment (richard67) [inline administrator/components/com_admin/sql/others/mysql/utf8mb4-conversion-02.sql]: @wilsonge Do you mean I shall make the username column same as the name column? Why so? It was not same before 3.5.

**Comment by richard67** [comment]
Removed reset of the conversion status from the schema update because this is 1. not helping in case of update method with file copy and database fix used, and 2. will be disturbing a soon to be developed solution for forcing conversion after an update.

So the former 2nd test method it not requred anymore, and so I have removed it from testing instructions. The remaining test which was former test method 1 is sufficient.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9510">issues.joomla.org/joomla-cms/9510</a>.</sub>

**Comment by richard67** [comment]
Ready for re-test.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9510">issues.joomla.org/joomla-cms/9510</a>.</sub>

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on b914f7ddd49b98cf6d82792b3a8b854bd4d174a0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9510">issues.joomla.org/joomla-cms/9510</a>.</sub>

**Comment by richard67** [comment]
Thanks for testing, @andrepereiradasilva

**Comment by richard67** [comment]
Closed in favour of #9549 .

**Pull request closed without merging**